api/agent/async.go

```diff
@@ -15,14 +16,15 @@ func (a *agent) asyncDequeue() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// parent span here so that we can see how many async calls are running
+	span, ctx := opentracing.StartSpanFromContext(ctx, "agent_async_dequeue")
+	defer span.Finish()
+
 	for {
-		ctxResource, cancelResource := context.WithCancel(context.Background())
 		select {
 		case <-a.shutdown:
-			cancelResource()
 			return
-		case <-a.resources.WaitAsyncResource(ctxResource):
-			cancelResource()
+		case <-a.resources.WaitAsyncResource(ctx):
```

Comment:
I think this is going to leak go routines. (See: resource go routine for cancel.)

Reply:
ping @rdallman

Reply:
I don't see anything, but maybe I'm missing it. In english:

select 1) we either are shutdown or we wait until async is ready
   * if we get a shutdown, return will call the deferred cancel to cancel the waiter's ctx
   * if we get a return from async resource it doesn't matter, we can reuse the ctx until we get a shutdown

select 2)
   * if we get a shutdown, return will call the deferred cancel to cancel asyncChew's ctx
   * if we get a message from asyncChew it doesn't matter, we can reuse the ctx until we get a shutdown
   * asyncRun only uses the ctx to extract the span

what'd I miss?

Reply:
right, inside of get resource token. inc.

Reply:
resolved (I think): https://github.com/fnproject/fn/pull/716/commits/e258e3ebc28e3c3bb1f27d8e33897cab8f62b477